src/svg/svg_parser.cpp

```diff
@@ -148,540 +164,497 @@ bool parse_style (const char* str, pairs_type & v)
-                using value_type = cont_type::value_type;
-                cont_type vec;
-                parse_style((const char*)value, vec);
-                for (value_type kv : vec )
-                {
-                    parse_attr(parser,BAD_CAST kv.first.c_str(),BAD_CAST kv.second.c_str());
-                }
+                parse_attr(parser, kv.first.c_str(), kv.second.c_str());
             }
-            else
-            {
-                parse_attr(parser,name,value);
-            }
-        } while(xmlTextReaderMoveToNextAttribute(reader) == 1);
+        }
+        else
+        {
+            parse_attr(parser,name, attr->value());
+        }
     }
-    xmlTextReaderMoveToElement(reader);
 }
 
-void parse_dimensions(svg_parser & parser, xmlTextReaderPtr reader)
+void parse_dimensions(svg_parser & parser, rapidxml::xml_node<char> const* node)
 {
-    xmlChar *value;
     double width = 0;
     double height = 0;
-    value = xmlTextReaderGetAttribute(reader, BAD_CAST "width");
-    if (value)
+    auto const* width_attr = node->first_attribute("width");
+    if (width_attr)
     {
-        width = parse_double((const char*)value);
-        xmlFree(value);
+        width = parse_double(parser.error_messages_, width_attr->value());
     }
-    xmlChar *value2;
-    value2 = xmlTextReaderGetAttribute(reader, BAD_CAST "width");
-    if (value2)
+    auto const* height_attr = node->first_attribute("height");
+    if (height_attr)
     {
-        height = parse_double((const char*)value2);
-        xmlFree(value2);
+        height = parse_double(parser.error_messages_, height_attr->value());
     }
-    parser.path_.set_dimensions(width,height);
-
+    parser.path_.set_dimensions(width, height);
 }
-void parse_path(svg_parser & parser, xmlTextReaderPtr reader)
-{
-    xmlChar *value;
 
-    value = xmlTextReaderGetAttribute(reader, BAD_CAST "d");
-    if (value)
+void parse_path(svg_parser & parser, rapidxml::xml_node<char> const* node)
+{
+    auto const* attr = node->first_attribute("d");
+    if (attr != nullptr)
     {
-        // d="" (empty paths) are valid
-        if (std::strlen((const char*)value) < 1)
-        {
-            xmlFree(value);
-        }
-        else
+        auto const* value = attr->value();
+        if (std::strlen(value) > 0)
         {
             parser.path_.begin_path();
 
-            if (!mapnik::svg::parse_path((const char*) value, parser.path_))
+            if (!mapnik::svg::parse_path(value, parser.path_))
             {
-                xmlFree(value);
-                xmlChar *id_value;
-                id_value = xmlTextReaderGetAttribute(reader, BAD_CAST "xml:id");
-                if (!id_value) id_value = xmlTextReaderGetAttribute(reader, BAD_CAST "id");
-                if (id_value)
+                auto const* id_attr = node->first_attribute("xml:id");
+                if (id_attr != nullptr) id_attr = node->first_attribute("id");
+                if (id_attr)
```

Comment:
Just to clarify, the way this code works now is that the path must have an `xml:id` property **and** and `id` property in order to have the error message print with the id. Is this the expected behavior? According to the [SVG spec](http://www.w3.org/TR/SVGMobile12/struct.html#idAttrs) it seems like there will only ever be `id` or `xml:id` on an element, and never both.

In that case, `if (id_attr != nullptr)` on line 493 should be `if (id_attr == nullptr)`, unless I'm reading the spec wrong.